logging_output_scripts/stat_analysis.py

```diff
@@ -0,0 +1,448 @@
+    # insights.
+    variants = ({
+        # Mean/median over cv runs per task (n_tasks problem instances).
+        "mean of":
+        lambda _df: _df[metric].groupby(["algorithm", "task"]).mean().unstack(
+        ).T,
+        "median of":
+        lambda _df: _df[metric].groupby(["algorithm", "task"]).median().
+        unstack().T,
+        # Each cv run as a separate problem instance (n_tasks * n_cv_runs problem
+        # instances).
+        "all":
+        lambda _df: _df[metric].unstack(0),
+    })
+    # Insight: We don't want to throw away distributional information.
+    if not all_variants:
+        variants = {"all": variants["all"]}
+
+    for metric in metrics:
+        # fig, ax = plt.subplots(len(variants), figsize=(linewidth, 2.7), dpi=72)
+        fig, ax = plt.subplots(
+            len(variants),
+            figsize=(textwidth,
+                     2.7 if metrics[metric] == "MSE" else 5 / 7 * 2.7),
+            dpi=72)
+        if not all_variants:
+            ax = [ax]
+        i = -1
+        for mode, f in variants.items():
+            i += 1
+
+            d = f(df)
+
+            # We want the algorithms ordered as they are in the `algorithms`
+            # list.
+            d = d[algorithms if not small_set else ["ES", "NS", "NSLC"]]
+
+            title = f"Considering {mode} cv runs per task"
+
+            print(
+                f"Sample statistics of {metrics[metric]} for “{title}” are as follows:"
+            )
+            print()
+            ranks = d.apply(np.argsort, axis=1) + 1
+            print(title)
+            smart_print(ranks.mean(), latex=latex)
+
+            # NOTE We fix the random seed here to enable model caching.
+            model = cmpbayes.Calvo(d.to_numpy(),
+                                   higher_better=False,
+                                   algorithm_labels=d.columns.to_list()).fit(
+                                       num_samples=10000, random_seed=1)
+
+            if check_mcmc:
+                smart_print(az.summary(model.data_), latex=latex)
+                az.plot_trace(model.data_)
+                az.plot_rank(model.data_)
+
+            # Join all chains, name columns.
+            sample = np.concatenate(model.data_.posterior.weights)
+            sample = pd.DataFrame(
+                sample, columns=model.data_.posterior.weights.algorithm_labels)
+            ylabel = "RD method"
+            # xlabel = f"Probability of having the lowest {metrics[metric]}"
+            xlabel = f"Probability"
+            sample = sample.unstack().reset_index(0).rename(columns={
+                "level_0": ylabel,
+                0: xlabel
+            })
+
+            if metrics[metric] == "MSE":
+                add_prob = sample[sample[ylabel] == "ES"][xlabel] + sample[
+                    sample[ylabel] == "NSLC"][xlabel]
+                add_prob = pd.DataFrame({
+                    ylabel:
+                    np.repeat("ES $\wedge$ NSLC", len(add_prob)),
```

Comment:
this is AND but should be OR

Reply:
https://github.com/dpaetzel/stat-analysis-rule-discovery-lcs/commit/c6cf8fd1b89b1a4238e8047028b66bbbf1856d5d